packages/client/src/publish/MessageFactory.ts

```diff
@@ -0,0 +1,105 @@
+import { random } from 'lodash'
+import { EncryptedGroupKey, EncryptionType, EthereumAddress, StreamID, StreamMessage, StreamPartID, toStreamPartID } from 'streamr-client-protocol'
+import { CacheConfig } from '../Config'
+import { EncryptionUtil } from '../encryption/EncryptionUtil'
+import { GroupKeyId } from '../encryption/GroupKey'
+import { CacheFn } from '../utils/caches'
+import { MessageChain } from './MessageChain'
+import { MessageMetadata } from './Publisher'
+import { keyToArrayIndex } from '@streamr/utils'
+import { GroupKeySequence } from './GroupKeyQueue'
+
+export interface MessageFactoryOptions {
+    streamId: StreamID
+    partitionCount: number
+    isPublicStream: boolean
+    publisherId: EthereumAddress
+    createSignature: (payload: string) => Promise<string>
+    useGroupKey: () => Promise<GroupKeySequence>
+    cacheConfig?: CacheConfig
+}
+
+export class MessageFactory {
+
+    private readonly streamId: StreamID
+    private readonly partitionCount: number
+    private readonly selectedDefaultPartition: number
+    private readonly isPublicStream: boolean
+    private readonly publisherId: EthereumAddress
+    private readonly createSignature: (payload: string) => Promise<string>
+    private readonly useGroupKey: () => Promise<GroupKeySequence>
+    private readonly getStreamPartitionForKey: (partitionKey: string | number) => number
+    private readonly getMsgChain: (streamPartId: StreamPartID, publisherId: EthereumAddress, msgChainId?: string) => MessageChain
+
+    constructor(opts: MessageFactoryOptions) {
+        this.streamId = opts.streamId
+        this.partitionCount = opts.partitionCount
+        this.selectedDefaultPartition = random(opts.partitionCount - 1)
+        this.isPublicStream = opts.isPublicStream
```

Comment:
We should not store these values here but instead rely on (cached) functions as the underlying facts can change in the smart contract between publishes.

Reply:
Fixed